driver/attributes.h

```diff
@@ -45,80 +44,85 @@ inline bool AttributeContainer::hasAttrAs<std::string>(int attr) const {
-inline std::string AttributeContainer::getAttrAs<std::string>(int attr, const std::string & def) const {
-    auto it = strings.find(attr);
-    return (it == strings.end() ?
-        def : it->second);
+template <typename T>
+inline void AttributeContainer::setAttrSilent(int attr, const T& value) {
+    if constexpr (std::is_pointer_v<T>)
+        attributes.insert_or_assign(attr, reinterpret_cast<std::intptr_t>(value));
+    else
+        attributes.insert_or_assign(attr, value);
 }
 
 template <typename T>
-inline T AttributeContainer::setAttrSilent(int attr, const T& value) {
-    std::int64_t old_value = 0;
-    strings.erase(attr);
-    auto it = integers.find(attr);
-    if (it == integers.end()) {
-        integers.emplace(attr, (std::int64_t)value);
-    }
-    else {
-        old_value = it->second;
-        it->second = (std::int64_t)value;
-    }
-    return (T)old_value;
-}
+inline void AttributeContainer::setAttr(int attr, const T& value) {
+    const auto it = attributes.find(attr);
 
-template <>
-inline std::string AttributeContainer::setAttrSilent<std::string>(int attr, const std::string& value) {
-    std::string old_value;
-    integers.erase(attr);
-    auto it = strings.find(attr);
-    if (it == strings.end()) {
-        strings.emplace(attr, value);
-    }
-    else {
-        old_value = std::move(it->second);
-        it->second = value;
-    }
-    return old_value;
-}
+    if (it == attributes.end()) {
+        if constexpr (std::is_pointer_v<T>)
+            attributes.emplace(attr, reinterpret_cast<std::intptr_t>(value));
+        else
+            attributes.emplace(attr, value);
 
-template <typename T>
-inline T AttributeContainer::setAttr(int attr, const T& value) {
-    std::int64_t old_value = 0;
-    strings.erase(attr);
-    auto it = integers.find(attr);
-    if (it == integers.end()) {
-        integers.emplace(attr, (std::int64_t)value);
         onAttrChange(attr);
     }
     else {
-        old_value = it->second;
-        it->second = (std::int64_t)value;
-        if (old_value != (std::int64_t)value)
-            onAttrChange(attr);
-    }
-    return (T)old_value;
-}
+        const bool changed = std::visit(
+            [&new_value = value] (auto & old_value) {
+                using NewValueType = std::decay_t<decltype(new_value)>;
+                using OldValueType = std::decay_t<decltype(old_value)>;
+
+                if constexpr (std::is_same_v<OldValueType, std::string>) {
+                    if constexpr (std::is_same_v<NewValueType, std::string>)
+                        return (old_value != new_value);
+                }
+                else {
+                    if constexpr (std::is_pointer_v<NewValueType>)
+                        return (old_value != reinterpret_cast<std::intptr_t>(new_value));
+                    else if constexpr (!std::is_same_v<NewValueType, std::string>)
+                        return (old_value != static_cast<std::intptr_t>(new_value));
+                }
```

Comment:
IMO, that should be simplified to something like:
```
const auto new_val = to_storage_type(new_value);
const auto old_val = to_storage_type(old_value);
if constexpr (std::is_same_v<decltype(new_val), decltype(old_val)>)
    return new_val != old_val;

// types are different, hence value is defiantly changed
return true;
```

and that `to_storage_type` function should have overloads/specializations to handle the pointer to int transformations, removing that from every other place.

Reply:
This is performance critical code, and any of types could be std::string. Creating temporary strings here is not something we want to do.

Also, hardly having lots of to_storage_type overloads/specializations will contribute readability. `if constexp` specifically helps to avoid exactly that.

Reply:
```
const std::string & to_storage_type(const std::string & value)
{
    return value;
}

// ...

const auto & new_val = to_storage_type(new_value);
```

The thing is that I see more than one case of explicit casting from pointer to `intptr_t`, and hence there is a chance of forgetting to add or to using it incorrectly. Also, that adds a perfect

Reply:
Any missing cast would cause compilation error.